Fix logic error in RenderLayer::willCompositeClipPath

rwlbuis opened the pull request:
#### 14afcfa777fec9ea51bed7d2f56c5bf786cea808
<pre>
Fix logic error in RenderLayer::willCompositeClipPath
<a href="https://bugs.webkit.org/show_bug.cgi?id=262648">https://bugs.webkit.org/show_bug.cgi?id=262648</a>

Reviewed by NOBODY (OOPS!).

Fix logic error in RenderLayer::willCompositeClipPath.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::willCompositeClipPath const):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14afcfa777fec9ea51bed7d2f56c5bf786cea808

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22294 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22493 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23368 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24193 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20632 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22543 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26755 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22822 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21642 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22529 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22230 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19345 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25048 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19280 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20201 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26451 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20297 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20432 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24312 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20962 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17756 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20421 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24650 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20959 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/330fb09093ee64a77b63d6e1791760e91dba2815. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/18642

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/cbe3e4c7e0b3c5d734c1adfb4468e80439924a62. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/18642

rwlbuis commented:
Unless I am blind, the statement seems to say:
return (A || !A) && B;

Obviously the first term is always true.

From the first time the code is introduced I am not sure though if this is meant:
return A || (!A && B);

or just

return B;

@smfr do you know/can you remember the intend of this code?

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -737,7 +737,7 @@ bool RenderLayer::willCompositeClipPath() const
         return false;
 
-    return (clipPath->type() != PathOperation::Shape || clipPath->type() == PathOperation::Shape) && GraphicsLayer::supportsLayerType(GraphicsLayer::Type::Shape);
+    return clipPath->type() != PathOperation::Shape || (clipPath->type() == PathOperation::Shape && GraphicsLayer::supportsLayerType(GraphicsLayer::Type::Shape));
```
Can't this be
```
return clipPath->type() != PathOperation::Shape || GraphicsLayer::supportsLayerType(GraphicsLayer::Type::Shape));
```

rwlbuis replied:
Done, my bigger point was though that I did not know which of the two ways stated above was the intention :)

rwlbuis committed:
Fix logic error in RenderLayer::willCompositeClipPath
https://bugs.webkit.org/show_bug.cgi?id=262648

Reviewed by NOBODY (OOPS!).

Fix logic error in RenderLayer::willCompositeClipPath.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::willCompositeClipPath const):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/14afcfa777fec9ea51bed7d2f56c5bf786cea808. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/18642

karlcow commented:
@rwlbuis by chance, does it progress some WPT tests in https://wpt.fyi/results/css/css-masking?label=master&label=experimental&product=chrome&product=firefox&product=safari&aligned&view=interop&q=label%3Ainterop-2023-cssmasking

rwlbuis commented:
> @rwlbuis by chance, does it progress some WPT tests in https://wpt.fyi/results/css/css-masking?label=master&label=experimental&product=chrome&product=firefox&product=safari&aligned&view=interop&q=label%3Ainterop-2023-cssmasking

Sadly, I see no change of behaviour for those tests.
